Add documentation and validation for CSSPseudoSelectors.json

nt1m opened the pull request:
#### 204b4f9d9d47a0120e702686d8f4777c6e66aef2
<pre>
Add documentation and validation for CSSPseudoSelectors.json
<a href="https://bugs.webkit.org/show_bug.cgi?id=267031">https://bugs.webkit.org/show_bug.cgi?id=267031</a>
<a href="https://rdar.apple.com/120407844">rdar://120407844</a>

Reviewed by Simon Fraser.

Add validation for:
- field types
- unknown fields
- &quot;argument&quot; field values
- ensuring no-one uses &quot;argument&quot; along with &quot;user-agent-part&quot;

Also add some documentation inline in CSSPseudoSelectors.json.
Documentation for each field is also enforced through the script.

* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/process-css-pseudo-selectors.py:
(InputValidator):
(InputValidator.__init__):
(InputValidator.validate_fields):

Canonical link: <a href="https://commits.webkit.org/272651@main">https://commits.webkit.org/272651@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0bcc553dfd2829e4922659a1dd1be5ce2f0a2bef

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34385 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35103 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/29422 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33405 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13636 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8480 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28952 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32973 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/9514 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/29079 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/8287 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/29020 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36439 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/29578 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/29443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34552 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/8559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/6511 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/32412 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/10229 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4199 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/9144 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -1,4 +1,29 @@
+            "user-agent-part": [
+                "For pseudo-elements only. Whether the pseudo-element maps to an element in an user agent shadow tree.",
+                "They internally map to `PseudoElement::UserAgentPart` or `PseudoElement::UserAgentPartLegacyAlias` for aliases."
+            ]
```
Depending on which PR lands first, this might also have to document user-agent-part-string. Otherwise we should add it to the other PR.

nt1m replied:
Will keep that in mind, thanks

annevk commented on Source/WebCore/css/CSSPseudoSelectors.json:
```diff
@@ -1,4 +1,29 @@
+            "comment": "Add a description of how the pseudo is used. Especially useful for non-standard or internal ones.",
+            "condition": "Compile-time `#if ENABLE()` condition for the feature.",
+            "settings-flag": "Settings flag queried from CSSSelectorParserContext. Note that the flag needs to be manually added there.",
+            "status": "Specifies the standardization and shipping state of the pseudo.",
```
Should this document the possible values?

nt1m replied:
Right now the only value that's used is non-standard, it'll probably require a bit more thought with engineering/evangelism if we ever want to add more values. This maybe should be changed to a "non-standard": true boolean if we're not planning to extend the value space.

annevk commented on Source/WebCore/css/process-css-pseudo-selectors.py:
```diff
@@ -741,6 +795,7 @@ def main():
     input_data = json.load(input_file)
     webcore_defines = [i.strip() for i in sys.argv[-1].split(' ')]
 
+    InputValidator(input_data)
```
Arguably the style checker should run this, but not sure how involved that is.

nt1m replied:
This will fail at build time like it does for CSSProperties.json. I think it's slightly nicer that way, and it's also easier that to have validation implemented in the same file as that code generation.

annevk reviewed (commented):
Really nice that we have this file now. Would have made my life easier and will undoubtedly help the next engineer looking to add pseudo-elements.

nt1m committed:
Add documentation and validation for CSSPseudoSelectors.json
https://bugs.webkit.org/show_bug.cgi?id=267031
rdar://120407844

Reviewed by Simon Fraser.

Add validation for:
- field types
- unknown fields
- "argument" field values
- ensuring no-one uses "argument" along with "user-agent-part"

Also add some documentation inline in CSSPseudoSelectors.json.
Documentation for each field is also enforced through the script.

* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/process-css-pseudo-selectors.py:
(InputValidator):
(InputValidator.__init__):
(InputValidator.validate_fields):

Canonical link: https://commits.webkit.org/272651@main
